drivers/flash/spi_nand.c

```diff
@@ -0,0 +1,1038 @@
+		ecc_steps = data->ecc_steps;
+		for (uint8_t i = 0 ; ecc_steps; ecc_steps--, i += data->ecc_bytes, p += data->ecc_size) {
+			memset(data->nbc.bch->input_data, 0x0, (1 << data->nbc.bch->m) / 8);
+			memcpy(data->nbc.bch->input_data + data->ecc_bytes, p, data->ecc_size);
+
+			int ret = bch_decode(data->nbc.bch, data->nbc.bch->input_data, (unsigned int *)(data->ecc_code + i));
+			if (ret < 0) {
+				LOG_ERR("Reading data failed");
+				return -ENODEV;
+			}
+			memcpy(p, data->nbc.bch->input_data + data->ecc_bytes, data->ecc_size);
+		}
+
+		memcpy(dest, data->page_buf + offset, read_bytes);
+
+		dest = (uint8_t *)dest + chunk;
+
+		addr = (addr + SPI_NAND_PAGE_OFFSET) & (~SPI_NAND_PAGE_MASK);
+		size -= chunk;
+	}
+
+out:	release_device(dev);
+
+	return ret;
+}
+
+static int spi_nand_read(const struct device *dev, off_t addr, void *dest,
+			size_t size)
+{
+	int ret = 0;
+	struct spi_nand_data *data = dev->data;
+
+	if (data->ecc_bits == 0) {
+		if (data->continuous_read) {
+			ret = spi_nand_read_cont(dev, addr, dest, size);
+		} else {
+			ret = spi_nand_read_normal(dev, addr, dest, size);
+		}
+	} else {
+		ret = spi_nand_read_software_ecc(dev, addr, dest, size);
+	}
+
+	return ret;
+}
+
+static int spi_nand_write(const struct device *dev, off_t addr,
+			 const void *src,
+			 size_t size)
+{
+	struct spi_nand_data *data = dev->data;
+	int ret = 0;
+	uint32_t offset = 0;
+	uint32_t chunk = 0;
+	uint32_t written_bytes = 0;
+
+	/* address must be sub-page-aligned */
+	if (!(addr & (SPI_NAND_SUB_PAGE_SIZE - 1)) && (addr != 0)) {
+		return -EINVAL;
+	}
+
+	/* size must be a multiple of sub-page */
+	if ((size % SPI_NAND_SUB_PAGE_SIZE) != 0) {
+		return -EINVAL;
+	}
+
+	acquire_device(dev);
+
+	while (size > 0) {
+
+		/* Don't write more than a page. */
+		offset = addr % data->page_size;
+		chunk = (offset + size < data->page_size) ? size : (data->page_size - offset);
+		written_bytes = chunk;
+
+		spi_nand_cmd_write(dev, SPI_NAND_CMD_WREN);
+
+		if (data->ecc_bits == 0) {
+			ret = spi_nand_access(dev, SPI_NAND_CMD_PP_LOAD,
+				NAND_ACCESS_WRITE | NAND_ACCESS_ADDRESSED | NAND_ACCESS_16BIT_ADDR,
+				(addr & SPI_NAND_PAGE_MASK), (void *)src, written_bytes);
+			if (ret != 0) {
+				LOG_ERR("program load failed: %d", ret);
+				goto out;
+			}
+		} else {
+			uint8_t *p = (uint8_t *)data->page_buf;
+			uint8_t ecc_steps = data->ecc_steps;
+
+			if (size < data->page_size) {
+				LOG_ERR("Write failed");
+				return -EINVAL;
+			}
+
+			/* prepare data */
+			memset(data->page_buf, 0xff, data->page_size + data->oob_size);
+			memcpy(data->page_buf + offset, (uint8_t *)src, written_bytes);
+
+			/* calculate the software ECC */
+			for (uint8_t i = 0; ecc_steps; ecc_steps--, i += data->ecc_bytes, p += data->ecc_size) {
+				memset(data->nbc.bch->input_data, 0x0, (1 << data->nbc.bch->m) / 8);
+				memcpy(data->nbc.bch->input_data + data->ecc_bytes, p, data->ecc_size);
+				bch_calculate_ecc(dev, data->nbc.bch->input_data, data->ecc_calc + i);
```

Comment:
Maybe we should not tie the ecc to bch and allow it to be passed via configured callback?

Reply:
Yes, Your suggestion is great, but currently we only implement BCH. Can we handle it this way in this version and wait until the next version implements other ECC algorithms before using the fallback function method？